clusterstate/clusterstate.go

```diff
@@ -0,0 +1,258 @@
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "api is down")
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "waiting for guest cluster")
+
+		err = c.guestFramework.WaitForGuestReady()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "guest cluster ready")
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "checking test app is installed")
+
+		err = c.CheckTestAppIsInstalled()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "test app is installed")
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "replacing master node")
+
+		err = c.provider.ReplaceMaster()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "master node replaced")
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "waiting api to go down")
+
+		err = c.guestFramework.WaitForAPIDown()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "api is down")
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "waiting for guest cluster")
+
+		err = c.guestFramework.WaitForGuestReady()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "guest cluster ready")
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "checking test app is installed")
+
+		err = c.CheckTestAppIsInstalled()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "test app is installed")
+	}
+
+	return nil
+}
+
+func (c *ClusterState) InstallTestApp() error {
+	var err error
+
+	var apprClient *apprclient.Client
+	{
+		c := apprclient.Config{
+			Fs:     afero.NewOsFs(),
+			Logger: c.logger,
+
+			Address:      CNRAddress,
+			Organization: CNROrganization,
+		}
+
+		apprClient, err = apprclient.New(c)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	var helmClient *helmclient.Client
+	{
+		c := helmclient.Config{
+			Logger:    c.logger,
+			K8sClient: c.guestFramework.K8sClient(),
+
+			RestConfig: c.guestFramework.RestConfig(),
+		}
+
+		helmClient, err = helmclient.New(c)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		err = helmClient.EnsureTillerInstalled()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	// Install the e2e app chart in the guest cluster.
+	{
+		c.logger.Log("level", "debug", "message", "installing e2e-app for testing")
+
+		tarballPath, err := apprClient.PullChartTarball(ChartName, ChartChannel)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		err = helmClient.InstallFromTarball(tarballPath, ChartNamespace)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	return nil
+}
+
+func (c *ClusterState) CheckTestAppIsInstalled() error {
```

Comment:
I followed the pattern used in e2e-harness. Maybe this could be moved to the framework. Could be useful if the label selector is passed as a param.